### ISSUE TITLE
[BUG] data value tooltip incorrect on figure.image with log scale

### ISSUE THREAD

**Issue by isaacto** [bokeh/bokeh#9663]
bokeh: 1.4.0
python: 3.7.5rc1
Jupyter: 1.0.0
OS: Linux, Ubuntu 19.10

I modified the example in Bokeh gallery called image.py so that one of the axis is in log scale.  While the figure is correct, the tooltips is not.  When hovering on a region with x being a multiple of pi, I expect to get a value close to 0.  But it usually does not.

My image map has x value between 1 to 10, in log scale.  In this range there should be 3 multiple of pi, where the value should be around 0.  I found that there are actually 3 vertical lines where the value is around 0, but they are at the wrong place: somewhere near x=5.5, 8.1 and 9.5, instead of the expected 3.1, 6.3 and 9.4.

So it looks to me that somewhere in the code the transformation bringing the screen coordinate to the axes coordinate is done twice incorrectly.

## Code that reproduces the issue

```
import numpy as np
from bokeh.plotting import show
from bokeh.models import ColorBar, LinearColorMapper

N = 500
x = np.logspace(0, 1, N)  # This is where the log scale data comes from
y = np.linspace(0, 10, N)
xx, yy = np.meshgrid(x, y)
d = np.sin(xx)*np.sin(yy)

# Here x-axis is configured to be log scale
p = figure(tooltips=[("x", "$x"), ("y", "$y"), ("value", "@image")], x_axis_type="log")
p.x_range.range_padding = p.y_range.range_padding = 0

p.image(image=[d], x=1, y=0, dw=9, dh=10, palette="Spectral11")
show(p)
```

#### Screenshot

![tooltip-bug](https://user-images.githubusercontent.com/1463929/74511243-8a005400-4f40-11ea-8837-99a8a62014e1.png)

**Comment by bryevdv** [comment]
@isaacto can you please experiment with other glyphs on a log axis and report back? We need to nail do whether this is a hover problem with log axes in general, or just somehow with images.

**Comment by isaacto** [comment]
Thanks for the response!  I've tried making a line graph instead, and it doesn't seem to be affected.  The "image" thing is quite different from other glyphs, and the input is 2D instead of 1D.  My guess is that the code supporting it needs to be quite different too, so the bug aren't the same.

~~~
import math
import numpy as np
import pandas as pd
from bokeh.plotting import show
from bokeh.models import ColorBar, LinearColorMapper
from bokeh.plotting import figure

N = 500
r = np.linspace(0, 2 * math.pi, N)
df = pd.DataFrame()
df['a'] = 3 + 2 * np.sin(r)
df['b'] = 3 + 2 * np.cos(r)
df['v'] = df['a'] + df['b']

p = figure(tooltips=[("x", "$x"), ("y", "$y"), ("a", "@a"), ("b", "@b"), ("v", "@v")],
           width=400, height=400, x_axis_type="log")
p.x_range.range_padding = p.y_range.range_padding = 0

p.line('a', 'b', source=df)
show(p)
~~~

![t](https://user-images.githubusercontent.com/1463929/74701781-def2e180-5242-11ea-9f06-58598dcbe701.png)

**Comment by bryevdv** [comment]
@isaacto Awesome, thank you for the additional example. I agree the issue seems to be specific to image plots. To be completely honest I am surprised (and don't really understand) how the log image is even rendering "correctly" to begin with. AFAIK there was never any code put in to make that happen, I didn't think anything related to log images was functional or supported (I would have actually expected the image to display "unstretched").

**Comment by isaacto** [comment]
Yes, I create the image that way.  Which is nothing new: when creating images for the linear axes users need to do the mapping from axes coordinates to screen coordinates (the (x, y) coordinate is at data[0][0], etc).  Indeed, when using logarithmic axes this is extremely important: otherwise the image will need to be orders of magnitude larger to make the small coordinate value side informative.

**Comment by bryevdv** [comment]
On `ImageBaseView`, fairly certain this needs to take the scale type into account
```typescript
  protected _image_index(index: number, x: number, y: number): ImageIndex {
    const [l, r, t, b] = this._lrtb(index)
    const width = this.image_width[index]
    const height = this.image_height[index]
    const dx = (r - l) / width
    const dy = (t - b) / height
    const i = Math.floor((x - l) / dx)
    const j = Math.floor((y - b) / dy)
    return {index, i, j, flat_index: j*width + i}
  }
```

**Comment by bryevdv** [comment]
e.g. this hard-coded computation to compute the index of the pixel offset in log space works as expected:
```typescript
const i = Math.floor(Math.log(x/l) / Math.log(r/l) * width)
```

<img width="540" alt="Image" src="https://github.com/user-attachments/assets/31b4f68b-3c1b-4ed1-b5e9-8d0c06e1b8f6" />

@mattpap what is the best way to interrogate the correct scale at this point? I can see e.g. `this.root.model.x_scale` is available, is there any better approach? It does not look like image glyphs can accept `x_range_name`, etc...

**Comment by mattpap** [comment]
In a glyph:
```ts
const {x_range, y_range} = this.renderer.coordinates
```
There are also `this.renderer.{xscale,yscale}` aliases.
